test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,225 @@
+	Byf("Writing a new clusterctl config to %s", modifiedClusterctlConfigPath)
+	copyAndAmendClusterctlConfig(ctx, copyAndAmendClusterctlConfigInput{
+		ClusterctlConfigPath: clusterctlConfigPath,
+		OutputPath:           modifiedClusterctlConfigPath,
+		Variables:            variables,
+	})
+
+	return modifiedClusterctlConfigPath, ipamService.cleanup
+}
+
+type ipamService struct {
+	client          client.Client
+	ipAddressClaims []ipamv1.IPAddressClaim
+	skipCleanup     bool
+}
+
+func newIPAMService(e2eIPAMKubeconfig string, skipCleanup bool) *ipamService {
+	kubeConfig, err := os.ReadFile(filepath.Clean(e2eIPAMKubeconfig))
+	Expect(err).ToNot(HaveOccurred())
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfig)
+	Expect(err).ToNot(HaveOccurred())
+
+	ipamClient, err := client.New(restConfig, client.Options{Scheme: ipamScheme})
+	Expect(err).ToNot(HaveOccurred())
+
+	return &ipamService{
+		ipAddressClaims: []ipamv1.IPAddressClaim{},
+		client:          ipamClient,
+		skipCleanup:     skipCleanup,
+	}
+}
+
+func (s *ipamService) claimIPAddress(ctx context.Context) string {
+	claim := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ipclaim-" + rand.String(32),
+			Namespace: metav1.NamespaceDefault,
+			Annotations: map[string]string{
+				"prow.k8s.io/build-id": os.Getenv("BUILD_ID"),
+				"prow.k8s.io/job":      os.Getenv("JOB_NAME"),
+			},
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				APIGroup: ptr.To("ipam.cluster.x-k8s.io"),
+				Kind:     "InClusterIPPool",
+				Name:     "capv-e2e-ippool",
+			},
+		},
+	}
+
+	// Create an IPAddressClaim
+	Byf("Creating IPAddressClaim %s/%s", claim.GetNamespace(), claim.GetName())
+	Expect(s.client.Create(ctx, claim)).To(Succeed())
+
+	// Wait for the IPAddressClaim to refer an IPAddress.
+	Eventually(func() error {
+		if err := s.client.Get(ctx, client.ObjectKeyFromObject(claim), claim); err != nil {
+			return err
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			return fmt.Errorf("IPAddressClaim does not yet have an IP")
+		}
+
+		return nil
+	}, time.Second*30, time.Second).Should(Succeed())
+
+	ip := &ipamv1.IPAddress{}
+
+	Expect(s.client.Get(ctx, client.ObjectKey{Namespace: claim.GetNamespace(), Name: claim.Status.AddressRef.Name}, ip)).To(Succeed())
+	Expect(ip.Spec.Address).ToNot(Equal(""))
+
+	// Store claim inside the service so the cleanup function knows what to delete.
+	s.ipAddressClaims = append(s.ipAddressClaims, *claim)
+
+	return ip.Spec.Address
+}
+
+// cleanup deletes the IPAddressClaims created by the service if the test succeeded.
+func (s *ipamService) cleanup() {
+	if s.skipCleanup {
+		By("Skipping cleanup of IPAddressClaims because skipCleanup is set")
+		return
+	}
+	if CurrentSpecReport().Failed() {
+		By("Skipping cleanup of IPAddressClaims because the tests failed and the IPs could still be in use")
+		return
+	}
```

Comment:
This looks like we will leak IPAM objects on all failed tests (?)

This seems pretty bad

Reply:
Hmm thinking about this again: yes that's bad 🤔 

However, we need to be careful, that an IP Address is really not in-use anymore (== Control Plane VMs do not exist anymore). Otherwise re-using that IP won't work either / cause issues.

Reply:
Yeah unfortunately. So we should only do this if the cluster has been deleted cleanly

Reply:
Implemented a Teardown in SynchronizedAfter which handles the deletion of these orphaned ones if possible.